source/volume_of_fluid/assembler.cc

```diff
@@ -0,0 +1,667 @@
+                                           face->boundary_id(),
+                                           scratch.face_finite_element_values.quadrature_point(q),
+                                           field.composition_index) *
+                                         current_u *
+                                         scratch.face_finite_element_values.normal_vector(q) *
+                                         scratch.face_finite_element_values.JxW(q);
+
+                }
+            }
+
+          // Due to inability to reference this cell's values at the interface,
+          // need to do explicit calculation
+          if (f_dir_pos)
+            {
+              face_ls_d = cell_i_d - 0.5*cell_i_normal[f_dim];
+              face_ls_time_grad = (face_flux/cell_vol)*cell_i_normal[f_dim];
+            }
+          else
+            {
+              face_ls_d = cell_i_d + 0.5*cell_i_normal[f_dim];
+              face_ls_time_grad = -(face_flux/cell_vol)*cell_i_normal[f_dim];
+            }
+
+          // Calculate outward flux
+          double flux_volume_of_fluid;
+          if (std::abs(face_flux) < volume_fraction_threshold*cell_vol)
+            {
+              flux_volume_of_fluid = cell_volume_of_fluid;
+            }
+          else if (face_flux < 0.0) // edge is upwind, currently assume zero inflow
+            {
+              flux_volume_of_fluid = boundary_fluid_flux/face_flux;
+            }
+          else // Cell is upwind, outflow boundary
+            {
+              flux_volume_of_fluid = VolumeOfFluid::calc_volume_of_fluid_flux_edge<dim> (f_dim,
+                                                                                         face_ls_time_grad,
+                                                                                         cell_i_normal,
+                                                                                         face_ls_d);
+            }
+
+          // Add fluxes to RHS
+          data.local_rhs[0] -= (flux_volume_of_fluid-cell_volume_of_fluid) * face_flux;
+        }
+    }
+
+    template <int dim>
+    void VolumeOfFluidAssembler<dim>::local_assemble_internal_face_volume_of_fluid_system (const VolumeOfFluidField<dim> field,
+        const unsigned int calc_dir,
+        bool update_from_old,
+        const typename DoFHandler<dim>::active_cell_iterator &cell,
+        const unsigned int face_no,
+        internal::Assembly::Scratch::VolumeOfFluidSystem<dim> &scratch,
+        internal::Assembly::CopyData::VolumeOfFluidSystem<dim> &data) const
+    {
+      const bool old_velocity_avail = (this->get_timestep_number() > 0);
+
+      const unsigned int f_dim = face_no/2; // Obtain dimension
+      const bool f_dir_pos = (face_no%2==1);
+
+      const unsigned int n_f_q_points    = scratch.face_finite_element_values.n_quadrature_points;
+
+      // vol fraction and interface values are constants, so can set from first value
+      const double cell_vol = cell->measure();
```

Comment:
As far as I remember `cell->measure()` is only an approximation of the cell volume. It might be exact for linear mappings though, so maybe it is find as is. Anyway, where do you need it, and could you easily replace it by a sum over `JxW(q)`, which gives you a better approximation for the real volume?